examples/handlers/stateful/src/main.rs

```diff
@@ -0,0 +1,119 @@
+//! An example of using stateful handlers with the Gotahm web framework.
+
+extern crate gotham;
+extern crate hyper;
+extern crate mime;
+
+use hyper::{Response, StatusCode};
+use std::sync::{Arc, Mutex};
+use std::time::SystemTime;
+
+use gotham::http::response::create_response;
+use gotham::router::Router;
+use gotham::router::builder::*;
+use gotham::state::State;
+
+// A struct which can store the state which it needs.
+struct CountingHandler {
+    // Record what time the server started.
+    // started_at will never change, so we can just store its value.
+    started_at: SystemTime,
+    // Count how many visits have been made to the server.
+    // visits will change each time the handler is called, so we need to wrap it an an Arc of a
+    // Mutex so that we can control concurrent access to it.
+    visits: Arc<Mutex<usize>>,
+}
+
+impl CountingHandler {
+    fn new() -> CountingHandler {
+        CountingHandler {
+            started_at: SystemTime::now(),
+            visits: Arc::new(Mutex::new(0)),
+        }
+    }
+
+    fn get_handler(&self, state: State) -> (State, Response) {
```

Comment:
Sorry for the delay on getting to look further at this PR.

I'd suggest implementing [Handler](https://docs.rs/gotham/0.2.1/gotham/handler/trait.Handler.html) for `CountingHandler` to do it "the Gotham way".

You'd then be able to simplify the router definition to:

`route.get("/").to(CountingHandler)`

(On the off chance you need it you could define [NewHandler](https://docs.rs/gotham/0.2.1/gotham/handler/trait.NewHandler.html) and then make use of `to_new_handler`).

Reply:
Thanks for taking a look!

Unfortunately I can't use `.to` because `CountingHandler` can't be `Clone`, because it contains non-Copyable state (the `Arc<Mutex<usize>>`).

I've updated the code to use `Handler` and `NewHandler`, but it makes me slightly sad because it forces an additional `clone()` of the handler per request (because `Handler.handle` takes `self` not `&self`) - not the end of the world, but something I feel should be avoidable. Is there a gotham-idiomatic way to both cheaply store state _and_ avoid having to copy/clone a `Handler` per request?

Reply:
Agreed that it isn't an optimal situation right now but necessitated.

@smangelsdorf and I will revisit this and see what is possible with tokio/futures/lifetimes as we adapt to the new direction Tokio is taking, likely as part of Hyper 0.12 work.